lib/src/spec/message.dart

```diff
@@ -49,38 +50,71 @@ class MessageData<T> {
   }
 }
 
+/// Delta extension configuration for [MessageExtras]
+class DeltaExtras {
+  /// the id of the message the delta was generated from
+  final String from;
+
+  /// the delta format. Only "vcdiff" is supported currently
+  final String format;
+
+  /// initializes with from and format
+  const DeltaExtras({this.from, this.format});
+
+  /// create instance from a map
+  static DeltaExtras _fromMap(Map value) {
+    if (value == null) return null;
+    return DeltaExtras(
+      from: value['from'] as String,
+      format: value['format'] as String,
+    );
+  }
+
+  @override
+  bool operator ==(Object other) =>
+      other is DeltaExtras && other.from == from && other.format == format;
+
+  @override
+  int get hashCode => '$from:$format'.hashCode;
+}
+
 /// Handles supported message extras types, their encoding and decoding
-class MessageExtras<T> {
-  final T _extras;
+class MessageExtras {
+  /// json-encodable map of extras
+  final Map<String, dynamic> extras;
+
+  /// configuration for delta compression extension
+  DeltaExtras _delta;
 
-  /// Only Map and List types are supported
-  MessageExtras(this._extras) : assert(T == Map);
+  /// delta configuration received from channel message
+  DeltaExtras get delta => _delta;
 
-  /// retrieve extras
-  T get extras => _extras;
+  /// extras is a mandatory json-encodable argument while delta is an extension
+  ///  which defines configuration for delta compression extension
+  MessageExtras(this.extras);
 
   /// initializes [MessageExtras] with given value and validates
   /// the data type, runtime
-  static MessageExtras fromValue(Object value) {
-    if (value == null) {
-      return null;
-    }
-    assert(
-      value is MessageExtras || value is Map,
-      'Message extras must be either `Map`, or `null`.'
-      ' Does not support $value ("${value.runtimeType}")',
-    );
-    if (value is MessageExtras) {
-      return value;
-    } else if (value is Map) {
-      return MessageExtras<Map>(value);
-    } else {
-      throw AssertionError(
-        'Message extras must be either `Map`, or `null`.'
-        ' Does not support $value ("${value.runtimeType}")',
-      );
-    }
+  static MessageExtras fromMap(Map<String, dynamic> value) {
+    if (value == null) return null;
+    value = Map<String, dynamic>.from(value);
+    final deltaMap = value.remove('delta') as Map;
+    final extras = MessageExtras(value)
+      .._delta = DeltaExtras._fromMap(deltaMap);
+    return extras;
   }
+
+  @override
+  String toString() => {'extras': extras, 'delta': delta}.toString();
+
+  @override
+  bool operator ==(Object other) =>
+      other is MessageExtras &&
+      const MapEquality().equals(other.extras, extras) &&
+      other.delta == delta;
+
+  @override
+  int get hashCode => '$extras:${delta.hashCode}'.hashCode;
```

Comment:
This is a very odd pattern (triggering potentially many `toString()` implementations), and I doubt particularly efficient. Can't you leverage the `hasCode` implementation from the `extras` map and combine that with `delta.hashCode`?

Reply:
9818326e39dad8cac0b12b87f333aac1a919c1f9

Reply:
Hmm, still not ideal as by "combine" I wasn't describing still using string representation to mix them together, however I recognise I wasn't clear on that. We need to get this PR landed so I'm not going to insist on this changing now but I have created an issue to improve in due course: https://github.com/ably/ably-flutter/issues/132